### ISSUE TITLE
SalesforceQuerySet update is failing

### ISSUE THREAD

**Issue by dhbahr** [django-salesforce/django-salesforce#198]
I'm trying to use the `SalesforceQueryset.update` method to set a Picklist value and I'm getting a TypeError, stating that what should be a `str` is in fact a `Query`...

```
  ...
  account_queryset.update(status=new_status)
  ...
```

```
TypeError at /bulk/
must be str, not Query
Request Method: | POST
```

these are the variables values when invoking `salesforce.backend.query.CursorWrapper.execute_update`

```
headers | {'Content-Type': 'application/json'}
pk | <django.db.models.sql.query.Query object at 0x7f10347774a8>
post_data | {'Status__c': 'NEW_STATUS'}
query | <django.db.models.sql.subqueries.UpdateQuery object at 0x7f103a63da20>
self | <salesforce.backend.query.CursorWrapper object at 0x7f1034777128>
table | 'Account'
```

and this is what gets through to the `rest_api_url` method itself

```
args | ('Account', <django.db.models.sql.query.Query object at 0x7f10347774a8>)
service | 'sobjects'
sf_session | <requests.sessions.Session object at 0x7f1034844d30>
```

**Comment by philchristensen** [comment]
Without more detailed code to be certain, it looks to me like you're accidentally passing a QuerySet object as `new_status` in the original call to update. That would be pretty easy to do accidentally.

What does the code look like before the first line you included? How do you define `new_status`?

**Comment by dhbahr** [comment]
I see what you mean, and I'm really sorry for the carelessness .. in this case I can assure you the value of `new_status` is "NEW_STATUS" as you can see in the `post_data` variable content

**Comment by hynekcer** [comment]
There is a bug that will be fixed soon. It works in some branches, it works with Django 1.10 or older but it is broken with 1.11. There is a workaround if you don't want to wait. You probably call `Account.objects.filter(pk__in=queryset).update(**something)`  
You can convert the internal queryset to list. (a real list, not `something.values_list(...)` because it is also a QuerySet not a list.)
That is Query instead of Queryset, probably by mistake.

What should be supported:
* `Model.objects.filter(pk__in=salesforce_queryset).update(**something)` (currently broken)
* `Model.objects.filter(pk__in=list_of_id).update(**something)`
* `Model.objects.filter(pk=some_id).update(**something)`

and the maximum number is 200 items.

This because SFDC conventions are to never update or delete many records unintentionally and to require explicit IDs before update or delete. We mimic it in the current implementation and we never increase the number of requests: to get IDs by one request and delete/update them by the second or eventually split them to more requests by 200, though the splitting is prepared for some new version.

If a query set is used, it should be Salesforce query set (based on SalesforceModel and stored in Salesforce database, otherwise you must even so convert the query set to list of IDs

**Comment by dhbahr** [comment]
It is a Salesforce QuerySet, @hynekcer, so that shouldn't be a problem. I'll test your workarounds ASAP and confirm here.
Many thanks!!!

**Comment by dhbahr** [comment]
@hynekcer so .. I'm actually trying to use a list of salesforce ids in the query set .. this is the variable context when invoking the queryset update:
```
form | <BulkActionsForm bound=True, valid=True, fields=(status;students;course)>
new_status | 'Abgebrochen'
request | <WSGIRequest: POST '/bulk/'>
self | <integration.views_staff.BulkActions object at 0x7f7ddd991908>
student_pks | ['0017E00000cYPovQAG',  '0017E00000cYPp1QAG',  '0017E00000cYPp6QAG',  '0017E00000cYPp8QAG',  '0017E00000cYPpAQAW',  '0017E00000cYPpFQAW']
students | <SalesforceQuerySet [<Account: Test Test6>, <Account: Test Test4>, <Account: Test Test8>, <Account: Test Test1>, <Account: Test Test9>, <Account: Test Test5>]>
```
and the relevant piece of code:

**Comment by dhbahr** [comment]
Well .. that was actually it .. if I use the standard SalesforceManager it works .. hope it serves to you as feedback .. feel free to close this one if you don't have any followup questions..

Happy hacking!

**Comment by hynekcer** [comment]
This is a useful issue report and I will fix it. (I forgot either to add it to README or to fix it soon when I had silenced it for Django 1.11 in test_bulk_update in April commit "e703649" in order to not postpone a release v0.7.0 with rough support for Django 1.11.)

Any custom manager for a SalesforceModel should be now a descendant of `SalesforceManager` (see README) primarily in order to use SalesforceQuerySet. This could be mitigated maybe next year in most cases because SalesforceQuerySet is used currently only for two special methods (query_all for searching everywhere including the trash bin, simple_select_related as replacement of select_related which is complicated to be fixed currently and select_related could be a little heavy weight for SFDC)

It is a good point to simplify the code later and remove the general requirement of use SalesforceManager. On the other hand it is a good hook that could be useful if a problem occurs with a new Django version and user's apps remain be ready for it. (This hacking is happy, anticipating and responsible.)

**Comment by dhbahr** [comment]
@hynekcer just to be completely clear: my custom managers on `SalesforceModel` models are actually subclasses of `SalesforceManager`.

**Comment by hynekcer** [comment]
I should note, that you don't need the filter condition `is_deleted=False` with a normal query set. The "trash bin" can be searched only after `query_all()` method.
